analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/NullPointerDereferenceBase.cs

```diff
@@ -20,34 +20,45 @@
 {
-    public abstract class NullPointerDereferenceBase : SymbolicRuleCheck
-    {
-        internal const string DiagnosticId = "S2259";
+    internal const string DiagnosticId = "S2259";
 
-        protected virtual bool IsSupressed(SyntaxNode node) => false;
+    protected virtual bool IsSupressed(SyntaxNode node) => false;
 
-        protected override ProgramState PreProcessSimple(SymbolicContext context)
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        if (ReferenceOrDefault(context.Operation.Instance) is { Syntax: var syntax} reference
+            && context.HasConstraint(reference, ObjectConstraint.Null)
+            && !IsSupressed(syntax)
+            && SemanticModel.GetTypeInfo(syntax).Nullability().FlowState != NullableFlowState.NotNull)
         {
-            var reference = context.Operation.Instance.Kind switch
-            {
-                OperationKindEx.Invocation => context.Operation.Instance.ToInvocation().Instance,
-                OperationKindEx.PropertyReference => context.Operation.Instance.ToPropertyReference().Instance,
-                OperationKindEx.Await => context.Operation.Instance.ToAwait().Operation,
-                OperationKindEx.ArrayElementReference => context.Operation.Instance.ToArrayElementReference().ArrayReference,
-                _ => null,
-            };
-            if (reference != null
-                && context.HasConstraint(reference, ObjectConstraint.Null)
-                && !reference.Type.IsStruct() // ToDo: IsStruct() is a workaround before MMF-2401
-                && !IsSupressed(reference.Syntax)
-                && SemanticModel.GetTypeInfo(reference.Syntax).Nullability().FlowState != NullableFlowState.NotNull)
-            {
-                ReportIssue(reference, reference.Syntax.ToString());
-            }
-
-            return context.State;
+            ReportIssue(reference, syntax.ToString());
         }
+        return context.State;
     }
+
+    private static IOperation ReferenceOrDefault(IOperation operation) =>
+        operation.Kind switch
+        {
+            OperationKindEx.Invocation => InvocationOrDefault(operation),
+            OperationKindEx.PropertyReference => PropertyReferenceOrDefault(operation),
+            OperationKindEx.Await => operation.ToAwait().Operation,
+            OperationKindEx.ArrayElementReference => operation.ToArrayElementReference().ArrayReference,
+            _ => null,
+        };
+
+    private static IOperation InvocationOrDefault(IOperation operation) =>
+        operation.ToInvocation() is { TargetMethod: var method } invocation
+        && !method.IsAny(KnownType.System_Nullable_T, nameof(Nullable<int>.GetValueOrDefault), nameof(Nullable<int>.Equals), nameof(Nullable<int>.ToString), nameof(Nullable<int>.GetHashCode))     // GetType raises, since it throws NRE
+            ? invocation.Instance
+            : null;
```

Comment:
No, I might use your trick.

Reply:
btw this `operation.ToInvocation() is { TargetMethod: var method } invocation` would be better like
`operation.ToInvocation() is var invocation` because 
* `ToInvocation` never returns null - it doesn't need the null check
* `method` is used just once, so there's no reason to rename it. You can use `invocation.TargetMethod.IsAny(...)` that is easier to read without `method` indirection